### PR TITLE
controller: enhance timeout error message for Recreate deployments

### PR DESCRIPTION
**Pull request by 0xmichalis**
Makes the error message from https://github.com/kubernetes/kubernetes/issues/29197 more obvious

@kubernetes/deployment

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/29505)

### CODE REVIEW

**Comment by kargakis** [comment]
@janetkuo can you have a look?

**Comment by kargakis** [comment]
@k8s-bot test this issue: #IGNORE

File: pkg/controller/deployment/recreate.go [hunk 1]
```diff
@@ -99,9 +102,27 @@ func (dc *DeploymentController) scaleDownOldReplicaSetsForRecreate(oldRSs []*ext
...
+		specReplicas := rs.Spec.Replicas
+		statusReplicas := rs.Status.Replicas
+
+		if err := wait.ExponentialBackoff(unversionedclient.DefaultRetry, func() (bool, error) {
+			replicaSet, err := dc.rsStore.ReplicaSets(rs.Namespace).Get(rs.Name)
+			if err != nil {
+				return false, err
+			}
+
+			specReplicas = replicaSet.Spec.Replicas
```
Comment (mfojtik) [inline pkg/controller/deployment/recreate.go]: do you need the spec.replicas here? isn't checking status.replicas enough?

**Review by mfojtik — APPROVED** [review]
LGTM

File: pkg/controller/deployment/recreate.go (hunk 1 above)
Comment (kargakis) [inline pkg/controller/deployment/recreate.go]: No, status.replicas=0 doesn't guarantee spec.replicas=0. Actually that means that the replica set controller is about to create new pods.

**Comment by kargakis** [comment]
Added TODO to use TerminatingReplicas (https://github.com/kubernetes/kubernetes/issues/32567), if we ever get that field. Applying lgtm

File: pkg/controller/deployment/recreate.go (hunk 1 above)
Comment (kargakis) [inline pkg/controller/deployment/recreate.go]: > Actually that means that the replica set controller is about to create new pods.

The case where status.replicas=0 and spec.replicas>0

**Pull request merged**